### ISSUE TITLE
Implement better session pooling.

### ISSUE THREAD

**Issue by coryan** [googleapis/google-cloud-cpp-spanner#307]
To unblock the implementation of `Client::Commit()` I implemented a very naive session pool. Here are some of the things it does not do:

- It does not remove sessions that have likely expired (i.e. unused for more than 1 hour).
- It does not automatically refresh the sessions in the background (consider using a `CompletionQueue` to do this with application-provided threads).
- It does not support a Stub that has multiple channels under the hood.
- It does not warm the session pool during the class initialization (consider doing this with background threads).

**Comment by mr-salty** [comment]
Assigning to myself to do some research / design work on this; even if we don't implement it (fully or at all) for alpha, we've run into several case where other decisions are predicated on having some understanding of how we eventually expect it to work.

**Comment by devjgm** [comment]
Any update on this? Looks like there's been lots of progress. Is this done?

**Comment by mr-salty** [comment]
not done, I have one more PR in progress and then I think I can declare it "done enough to close this bug" and open issues for any small things leftover (in retrospect I probably should have opened multiple issues at the start instead of using this for everything)

**Comment by devjgm** [comment]
Can we file a separate issue that clearly describes what remaining feature/functionality we need to add in this next PR you're referring to? I'm thinking then we can probably label *that* issue with GA (not beta), then close this one for Beta.

FTR, this is related to what Todd and I talked about offline.

**Comment by mr-salty** [comment]
The 4 bullets I listed initially probably should have been separate issues.

3 of them are done: utilizing multiple channels, proactively creating `min_sessions`, and removing expired `Sessions` from the pool (which had already been split out to #914).

I opened #1171 for `Session` refresh and #1172 as an additional issue that wasn't identified here (properly maintaining `min_sessions_ and `max_idle_sessions`) - but as we discussed we are comfortable going beta without those implemented, so I'm closing this issue.
